File: packages/stdlib-core/implementations/typescript/ids.ts
// ============================================================================
// ISL Standard Library - TypeScript Identifier Implementation
// @stdlib/ids
// ============================================================================

// ============================================================================
// BRANDED TYPES
// ============================================================================

declare const __brand: unique symbol;
type Brand<T, B> = T & { [__brand]: B };

export type UUID = Brand<string, 'UUID'>;
export type UUIDv7 = Brand<string, 'UUIDv7'>;
export type CompactUUID = Brand<string, 'CompactUUID'>;
export type ULID = Brand<string, 'ULID'>;
export type KSUID = Brand<string, 'KSUID'>;
export type NanoID = Brand<string, 'NanoID'>;
export type ShortId = Brand<string, 'ShortId'>;
export type ObjectId = Brand<string, 'ObjectId'>;
export type SnowflakeId = Brand<string, 'SnowflakeId'>;
export type Cursor = Brand<string, 'Cursor'>;

// ============================================================================
// REGEX PATTERNS
// ============================================================================

export const ID_PATTERNS = {
  UUID_V4: /^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i,
  UUID_V7: /^[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i,
  UUID_ANY: /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i,
  COMPACT_UUID: /^[0-9a-f]{32}$/i,
  ULID: /^[0123456789ABCDEFGHJKMNPQRSTVWXYZ]{26}$/i,
  KSUID: /^[0-9A-Za-z]{27}$/,
  NANO_ID: /^[A-Za-z0-9_-]{21}$/,
  SHORT_ID: /^[A-Za-z0-9_-]{8,12}$/,
  HUMAN_CODE: /^[23456789ABCDEFGHJKLMNPQRSTUVWXYZ]{6}$/,
  OBJECT_ID: /^[0-9a-f]{24}$/,
  SNOWFLAKE: /^\d{18,19}$/,
  EAN13: /^\d{13}$/,
  UPC_A: /^\d{12}$/,
  ISBN13: /^97[89]\d{10}$/,
  ISBN10: /^\d{9}[\dX]$/,
  DOI: /^10\.\d{4,9}\/[-._;()\/:A-Z0-9]+$/i,
  ORCID: /^\d{4}-\d{4}-\d{4}-\d{3}[\dX]$/,
  STRIPE_CUSTOMER: /^cus_[A-Za-z0-9]{14,}$/,
  STRIPE_PAYMENT_INTENT: /^pi_[A-Za-z0-9]{24,}$/,
  STRIPE_SUBSCRIPTION: /^sub_[A-Za-z0-9]{14,}$/,
  ARN: /^arn:aws:[a-z0-9-]+:[a-z0-9-]*:\d{12}:.+$/,
  GITHUB_REPO: /^[a-zA-Z0-9](?:[a-zA-Z0-9]|-(?=[a-zA-Z0-9])){0,38}\/[a-zA-Z0-9._-]{1,100}$/,
  K8S_NAME: /^[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?$/,
  API_KEY: /^(sk|pk)_(live|test)_[A-Za-z0-9]{32,}$/,
} as const;

// ============================================================================
// VALIDATION FUNCTIONS
// ============================================================================

export function isValidUUID(value: string): value is UUID {
  return ID_PATTERNS.UUID_V4.test(value);
}

export function isValidUUIDv7(value: string): value is UUIDv7 {
  return ID_PATTERNS.UUID_V7.test(value);
}

export function isValidUUIDAny(value: string): boolean {
  return ID_PATTERNS.UUID_ANY.test(value);
}

export function isValidCompactUUID(value: string): value is CompactUUID {
  return ID_PATTERNS.COMPACT_UUID.test(value);
}

export function isValidULID(value: string): value is ULID {
  return ID_PATTERNS.ULID.test(value);
}

export function isValidKSUID(value: string): value is KSUID {
  return ID_PATTERNS.KSUID.test(value);
}

export function isValidNanoID(value: string): value is NanoID {
  return ID_PATTERNS.NANO_ID.test(value);
}

export function isValidShortId(value: string): value is ShortId {
  return ID_PATTERNS.SHORT_ID.test(value);
}

export function isValidHumanCode(value: string): boolean {
  return ID_PATTERNS.HUMAN_CODE.test(value);
}

export function isValidObjectId(value: string): value is ObjectId {
  return ID_PATTERNS.OBJECT_ID.test(value);
}

export function isValidSnowflakeId(value: string): value is SnowflakeId {
  return ID_PATTERNS.SNOWFLAKE.test(value);
}

// ============================================================================
// BARCODE / PRODUCT ID VALIDATION
// ============================================================================

export function isValidEAN13(value: string): boolean {
  if (!ID_PATTERNS.EAN13.test(value)) return false;
  return validateEAN13Checksum(value);
}

export function isValidUPCA(value: string): boolean {
  if (!ID_PATTERNS.UPC_A.test(value)) return false;
  return validateUPCAChecksum(value);
}

export function isValidISBN13(value: string): boolean {
  if (!ID_PATTERNS.ISBN13.test(value)) return false;
  return validateISBN13Checksum(value);
}

export function isValidISBN10(value: string): boolean {
  if (!ID_PATTERNS.ISBN10.test(value)) return false;
  return validateISBN10Checksum(value);
}

function validateEAN13Checksum(code: string): boolean {
  let sum = 0;
  for (let i = 0; i < 12; i++) {
    sum += parseInt(code[i]) * (i % 2 === 0 ? 1 : 3);
  }
  const checkDigit = (10 - (sum % 10)) % 10;
  return checkDigit === parseInt(code[12]);
}

function validateUPCAChecksum(code: string): boolean {
  let sum = 0;
  for (let i = 0; i < 11; i++) {
    sum += parseInt(code[i]) * (i % 2 === 0 ? 3 : 1);
  }
  const checkDigit = (10 - (sum % 10)) % 10;
  return checkDigit === parseInt(code[11]);
}

function validateISBN13Checksum(code: string): boolean {
  return validateEAN13Checksum(code);
}

function validateISBN10Checksum(code: string): boolean {
  let sum = 0;
  for (let i = 0; i < 9; i++) {
    sum += parseInt(code[i]) * (10 - i);
  }
  const lastChar = code[9];
  const lastValue = lastChar === 'X' ? 10 : parseInt(lastChar);
  sum += lastValue;
  return sum % 11 === 0;
}

// ============================================================================
// SERVICE-SPECIFIC VALIDATORS
// ============================================================================

export function isValidDOI(value: string): boolean {
  return ID_PATTERNS.DOI.test(value);
}

export function isValidORCID(value: string): boolean {
  if (!ID_PATTERNS.ORCID.test(value)) return false;
  return validateORCIDChecksum(value);
}

function validateORCIDChecksum(orcid: string): boolean {
  const digits = orcid.replace(/-/g, '');
  let total = 0;
  for (let i = 0; i < 15; i++) {
    total = (total + parseInt(digits[i])) * 2;
  }
  const remainder = total % 11;
  const checkDigit = (12 - remainder) % 11;
  const expected = checkDigit === 10 ? 'X' : checkDigit.toString();
  return digits[15] === expected;
}

export function isValidStripeCustomerId(value: string): boolean {
  return ID_PATTERNS.STRIPE_CUSTOMER.test(value);
}

export function isValidStripePaymentIntentId(value: string): boolean {
  return ID_PATTERNS.STRIPE_PAYMENT_INTENT.test(value);
}

export function isValidStripeSubscriptionId(value: string): boolean {
  return ID_PATTERNS.STRIPE_SUBSCRIPTION.test(value);
}

export function isValidARN(value: string): boolean {
  return ID_PATTERNS.ARN.test(value);
}

export function isValidGitHubRepo(value: string): boolean {
  return ID_PATTERNS.GITHUB_REPO.test(value);
}

export function isValidK8sName(value: string): boolean {
  return ID_PATTERNS.K8S_NAME.test(value);
}

export function isValidAPIKey(value: string): boolean {
  return ID_PATTERNS.API_KEY.test(value);
}

// ============================================================================
// UUID GENERATION (v4)
// ============================================================================

export function generateUUID(): UUID {
  // Use crypto.randomUUID if available (Node.js 19+ / modern browsers)
  if (typeof crypto !== 'undefined' && crypto.randomUUID) {
    return crypto.randomUUID() as UUID;
  }
  
  // Fallback implementation
  const bytes = new Uint8Array(16);
  if (typeof crypto !== 'undefined' && crypto.getRandomValues) {
    crypto.getRandomValues(bytes);
  } else {
    for (let i = 0; i < 16; i++) {
      bytes[i] = Math.floor(Math.random() * 256);
    }
  }
  
  // Set version (4) and variant (RFC4122)
  bytes[6] = (bytes[6] & 0x0f) | 0x40;
  bytes[8] = (bytes[8] & 0x3f) | 0x80;
  
  const hex = Array.from(bytes).map(b => b.toString(16).padStart(2, '0')).join('');
  return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20)}` as UUID;
}

// ============================================================================
// ULID GENERATION
// ============================================================================

const ULID_ENCODING = '0123456789ABCDEFGHJKMNPQRSTVWXYZ';

export function generateULID(): ULID {
  const timestamp = Date.now();
  const randomBytes = new Uint8Array(10);
  
  if (typeof crypto !== 'undefined' && crypto.getRandomValues) {
    crypto.getRandomValues(randomBytes);
  } else {
    for (let i = 0; i < 10; i++) {
      randomBytes[i] = Math.floor(Math.random() * 256);
    }
  }
  
  // Encode timestamp (first 10 characters)
  let timestampPart = '';
  let t = timestamp;
  for (let i = 0; i < 10; i++) {
    timestampPart = ULID_ENCODING[t % 32] + timestampPart;
    t = Math.floor(t / 32);
  }
  
  // Encode randomness (last 16 characters)
  let randomPart = '';
  for (let i = 0; i < 10; i++) {
    randomPart += ULID_ENCODING[randomBytes[i] % 32];
    if (i % 2 === 1 && i < 9) {
      randomPart += ULID_ENCODING[Math.floor(randomBytes[i] / 32)];
    }
  }
  randomPart = randomPart.slice(0, 16);
  
  return (timestampPart + randomPart) as ULID;
}

// ============================================================================
// SHORT ID GENERATION
// ============================================================================

const SHORT_ID_CHARS = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_-';
const HUMAN_CODE_CHARS = '23456789ABCDEFGHJKLMNPQRSTUVWXYZ';

export function generateShortId(length: number = 10): ShortId {
  if (length < 8 || length > 12) {
    throw new Error('ShortId length must be between 8 and 12');
  }
  
  let result = '';
  const bytes = new Uint8Array(length);
  
  if (typeof crypto !== 'undefined' && crypto.getRandomValues) {
    crypto.getRandomValues(bytes);
  } else {
    for (let i = 0; i < length; i++) {
      bytes[i] = Math.floor(Math.random() * 256);
    }
  }
  
  for (let i = 0; i < length; i++) {
    result += SHORT_ID_CHARS[bytes[i] % SHORT_ID_CHARS.length];
  }
  
  return result as ShortId;
}

export function generateHumanCode(length: number = 6): string {
  let result = '';
  const bytes = new Uint8Array(length);
  
  if (typeof crypto !== 'undefined' && crypto.getRandomValues) {
    crypto.getRandomValues(bytes);
  } else {
    for (let i = 0; i < length; i++) {
      bytes[i] = Math.floor(Math.random() * 256);
    }
  }
  
  for (let i = 0; i < length; i++) {
    result += HUMAN_CODE_CHARS[bytes[i] % HUMAN_CODE_CHARS.length];
  }
  
  return result;
}

// ============================================================================
// UUID UTILITIES
// ============================================================================

export function uuidToCompact(uuid: UUID): CompactUUID {
  return uuid.replace(/-/g, '').toLowerCase() as CompactUUID;
}

export function compactToUUID(compact: CompactUUID): UUID {
  const hex = compact.toLowerCase();
  return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20)}` as UUID;
}

export function uuidToBytes(uuid: UUID): Uint8Array {
  const hex = uuid.replace(/-/g, '');
  const bytes = new Uint8Array(16);
  for (let i = 0; i < 16; i++) {
    bytes[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
  }
  return bytes;
}

export function bytesToUUID(bytes: Uint8Array): UUID {
  if (bytes.length !== 16) {
    throw new Error('UUID must be 16 bytes');
  }
  const hex = Array.from(bytes).map(b => b.toString(16).padStart(2, '0')).join('');
  return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20)}` as UUID;
}

// ============================================================================
// ULID UTILITIES
// ============================================================================

export function ulidToTimestamp(ulid: ULID): number {
  let timestamp = 0;
  for (let i = 0; i < 10; i++) {
    timestamp = timestamp * 32 + ULID_ENCODING.indexOf(ulid[i].toUpperCase());
  }
  return timestamp;
}

export function ulidToDate(ulid: ULID): Date {
  return new Date(ulidToTimestamp(ulid));
}

// ============================================================================
// SNOWFLAKE ID UTILITIES
// ============================================================================

// Discord-style snowflake epoch (2015-01-01)
const SNOWFLAKE_EPOCH = 1420070400000n;

export function snowflakeToTimestamp(snowflake: SnowflakeId): number {
  const id = BigInt(snowflake);
  const timestamp = Number((id >> 22n) + SNOWFLAKE_EPOCH);
  return timestamp;
}

export function snowflakeToDate(snowflake: SnowflakeId): Date {
  return new Date(snowflakeToTimestamp(snowflake));
}

// ============================================================================
// EXPORTS
// ============================================================================

export const Ids = {
  // Validation
  isValidUUID,
  isValidUUIDv7,
  isValidUUIDAny,
  isValidCompactUUID,
  isValidULID,
  isValidKSUID,
  isValidNanoID,
  isValidShortId,
  isValidHumanCode,
  isValidObjectId,
  isValidSnowflakeId,
  isValidEAN13,
  isValidUPCA,
  isValidISBN13,
  isValidISBN10,
  isValidDOI,
  isValidORCID,
  isValidStripeCustomerId,
  isValidStripePaymentIntentId,
  isValidStripeSubscriptionId,
  isValidARN,
  isValidGitHubRepo,
  isValidK8sName,
  isValidAPIKey,
  
  // Generation
  generateUUID,
  generateULID,
  generateShortId,
  generateHumanCode,
  
  // UUID utilities
  uuidToCompact,
  compactToUUID,
  uuidToBytes,
  bytesToUUID,
  
  // ULID utilities
  ulidToTimestamp,
  ulidToDate,
  
  // Snowflake utilities
  snowflakeToTimestamp,
  snowflakeToDate,
  
  // Constants
  ID_PATTERNS,
};

export default Ids;

File: packages/stdlib-core/implementations/python/ids.py
"""
ISL Standard Library - Python Identifier Implementation
@stdlib/ids
"""

import os
import re
import secrets
import time
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import NewType, Optional


# ============================================================================
# BRANDED TYPES
# ============================================================================

UUID = NewType("UUID", str)
UUIDv7 = NewType("UUIDv7", str)
CompactUUID = NewType("CompactUUID", str)
ULID = NewType("ULID", str)
KSUID = NewType("KSUID", str)
NanoID = NewType("NanoID", str)
ShortId = NewType("ShortId", str)
ObjectId = NewType("ObjectId", str)
SnowflakeId = NewType("SnowflakeId", str)


# ============================================================================
# REGEX PATTERNS
# ============================================================================

ID_PATTERNS = {
    "UUID_V4": re.compile(
        r"^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$",
        re.IGNORECASE,
    ),
    "UUID_V7": re.compile(
        r"^[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$",
        re.IGNORECASE,
    ),
    "UUID_ANY": re.compile(
        r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$",
        re.IGNORECASE,
    ),
    "COMPACT_UUID": re.compile(r"^[0-9a-f]{32}$", re.IGNORECASE),
    "ULID": re.compile(r"^[0123456789ABCDEFGHJKMNPQRSTVWXYZ]{26}$", re.IGNORECASE),
    "KSUID": re.compile(r"^[0-9A-Za-z]{27}$"),
    "NANO_ID": re.compile(r"^[A-Za-z0-9_-]{21}$"),
    "SHORT_ID": re.compile(r"^[A-Za-z0-9_-]{8,12}$"),
    "HUMAN_CODE": re.compile(r"^[23456789ABCDEFGHJKLMNPQRSTUVWXYZ]{6}$"),
    "OBJECT_ID": re.compile(r"^[0-9a-f]{24}$"),
    "SNOWFLAKE": re.compile(r"^\d{18,19}$"),
    "EAN13": re.compile(r"^\d{13}$"),
    "UPC_A": re.compile(r"^\d{12}$"),
    "ISBN13": re.compile(r"^97[89]\d{10}$"),
    "ISBN10": re.compile(r"^\d{9}[\dX]$"),
    "DOI": re.compile(r"^10\.\d{4,9}\/[-._;()\/:A-Z0-9]+$", re.IGNORECASE),
    "ORCID": re.compile(r"^\d{4}-\d{4}-\d{4}-\d{3}[\dX]$"),
    "STRIPE_CUSTOMER": re.compile(r"^cus_[A-Za-z0-9]{14,}$"),
    "STRIPE_PAYMENT_INTENT": re.compile(r"^pi_[A-Za-z0-9]{24,}$"),
    "STRIPE_SUBSCRIPTION": re.compile(r"^sub_[A-Za-z0-9]{14,}$"),
    "ARN": re.compile(r"^arn:aws:[a-z0-9-]+:[a-z0-9-]*:\d{12}:.+$"),
    "GITHUB_REPO": re.compile(
        r"^[a-zA-Z0-9](?:[a-zA-Z0-9]|-(?=[a-zA-Z0-9])){0,38}\/[a-zA-Z0-9._-]{1,100}$"
    ),
    "K8S_NAME": re.compile(r"^[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?$"),
    "API_KEY": re.compile(r"^(sk|pk)_(live|test)_[A-Za-z0-9]{32,}$"),
}


# ============================================================================
# VALIDATION FUNCTIONS
# ============================================================================


def is_valid_uuid(value: str) -> bool:
    """Validate UUID v4 format."""
    return bool(ID_PATTERNS["UUID_V4"].match(value))


def is_valid_uuid_v7(value: str) -> bool:
    """Validate UUID v7 format."""
    return bool(ID_PATTERNS["UUID_V7"].match(value))


def is_valid_uuid_any(value: str) -> bool:
    """Validate any UUID format."""
    return bool(ID_PATTERNS["UUID_ANY"].match(value))


def is_valid_compact_uuid(value: str) -> bool:
    """Validate compact UUID format (no dashes)."""
    return bool(ID_PATTERNS["COMPACT_UUID"].match(value))


def is_valid_ulid(value: str) -> bool:
    """Validate ULID format."""
    return bool(ID_PATTERNS["ULID"].match(value))


def is_valid_ksuid(value: str) -> bool:
    """Validate KSUID format."""
    return bool(ID_PATTERNS["KSUID"].match(value))


def is_valid_nano_id(value: str) -> bool:
    """Validate NanoID format."""
    return bool(ID_PATTERNS["NANO_ID"].match(value))


def is_valid_short_id(value: str) -> bool:
    """Validate short ID format."""
    return bool(ID_PATTERNS["SHORT_ID"].match(value))


def is_valid_human_code(value: str) -> bool:
    """Validate human-readable code format."""
    return bool(ID_PATTERNS["HUMAN_CODE"].match(value))


def is_valid_object_id(value: str) -> bool:
    """Validate MongoDB ObjectId format."""
    return bool(ID_PATTERNS["OBJECT_ID"].match(value))


def is_valid_snowflake_id(value: str) -> bool:
    """Validate Snowflake ID format."""
    return bool(ID_PATTERNS["SNOWFLAKE"].match(value))


# ============================================================================
# BARCODE / PRODUCT ID VALIDATION
# ============================================================================


def _validate_ean13_checksum(code: str) -> bool:
    """Validate EAN-13 checksum."""
    total = sum(
        int(code[i]) * (1 if i % 2 == 0 else 3) for i in range(12)
    )
    check_digit = (10 - (total % 10)) % 10
    return check_digit == int(code[12])


def _validate_upc_a_checksum(code: str) -> bool:
    """Validate UPC-A checksum."""
    total = sum(
        int(code[i]) * (3 if i % 2 == 0 else 1) for i in range(11)
    )
    check_digit = (10 - (total % 10)) % 10
    return check_digit == int(code[11])


def _validate_isbn10_checksum(code: str) -> bool:
    """Validate ISBN-10 checksum."""
    total = sum(int(code[i]) * (10 - i) for i in range(9))
    last_char = code[9]
    last_value = 10 if last_char == "X" else int(last_char)
    total += last_value
    return total % 11 == 0


def _validate_orcid_checksum(orcid: str) -> bool:
    """Validate ORCID checksum."""
    digits = orcid.replace("-", "")
    total = 0
    for i in range(15):
        total = (total + int(digits[i])) * 2
    remainder = total % 11
    check_digit = (12 - remainder) % 11
    expected = "X" if check_digit == 10 else str(check_digit)
    return digits[15] == expected


def is_valid_ean13(value: str) -> bool:
    """Validate EAN-13 barcode."""
    if not ID_PATTERNS["EAN13"].match(value):
        return False
    return _validate_ean13_checksum(value)


def is_valid_upc_a(value: str) -> bool:
    """Validate UPC-A barcode."""
    if not ID_PATTERNS["UPC_A"].match(value):
        return False
    return _validate_upc_a_checksum(value)


def is_valid_isbn13(value: str) -> bool:
    """Validate ISBN-13."""
    if not ID_PATTERNS["ISBN13"].match(value):
        return False
    return _validate_ean13_checksum(value)


def is_valid_isbn10(value: str) -> bool:
    """Validate ISBN-10."""
    if not ID_PATTERNS["ISBN10"].match(value):
        return False
    return _validate_isbn10_checksum(value)


def is_valid_doi(value: str) -> bool:
    """Validate DOI."""
    return bool(ID_PATTERNS["DOI"].match(value))


def is_valid_orcid(value: str) -> bool:
    """Validate ORCID."""
    if not ID_PATTERNS["ORCID"].match(value):
        return False
    return _validate_orcid_checksum(value)


# ============================================================================
# SERVICE-SPECIFIC VALIDATORS
# ============================================================================


def is_valid_stripe_customer_id(value: str) -> bool:
    """Validate Stripe customer ID."""
    return bool(ID_PATTERNS["STRIPE_CUSTOMER"].match(value))


def is_valid_stripe_payment_intent_id(value: str) -> bool:
    """Validate Stripe payment intent ID."""
    return bool(ID_PATTERNS["STRIPE_PAYMENT_INTENT"].match(value))


def is_valid_stripe_subscription_id(value: str) -> bool:
    """Validate Stripe subscription ID."""
    return bool(ID_PATTERNS["STRIPE_SUBSCRIPTION"].match(value))


def is_valid_arn(value: str) -> bool:
    """Validate AWS ARN."""
    return bool(ID_PATTERNS["ARN"].match(value))


def is_valid_github_repo(value: str) -> bool:
    """Validate GitHub repo identifier."""
    return bool(ID_PATTERNS["GITHUB_REPO"].match(value))


def is_valid_k8s_name(value: str) -> bool:
    """Validate Kubernetes resource name."""
    return bool(ID_PATTERNS["K8S_NAME"].match(value))


def is_valid_api_key(value: str) -> bool:
    """Validate API key format."""
    return bool(ID_PATTERNS["API_KEY"].match(value))


# ============================================================================
# UUID GENERATION
# ============================================================================


def generate_uuid() -> UUID:
    """Generate UUID v4."""
    import uuid as uuid_module

    return UUID(str(uuid_module.uuid4()))


# ============================================================================
# ULID GENERATION
# ============================================================================

ULID_ENCODING = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"


def generate_ulid() -> ULID:
    """Generate ULID."""
    timestamp = int(time.time() * 1000)

    # Encode timestamp (first 10 characters)
    timestamp_part = ""
    t = timestamp
    for _ in range(10):
        timestamp_part = ULID_ENCODING[t % 32] + timestamp_part
        t //= 32

    # Generate random bytes for randomness part
    random_bytes = secrets.token_bytes(10)
    random_part = ""
    for b in random_bytes:
        random_part += ULID_ENCODING[b % 32]
    random_part = random_part[:16]

    return ULID(timestamp_part + random_part)


# ============================================================================
# SHORT ID GENERATION
# ============================================================================

SHORT_ID_CHARS = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_-"
HUMAN_CODE_CHARS = "23456789ABCDEFGHJKLMNPQRSTUVWXYZ"


def generate_short_id(length: int = 10) -> ShortId:
    """Generate short ID."""
    if not 8 <= length <= 12:
        raise ValueError("ShortId length must be between 8 and 12")

    return ShortId("".join(secrets.choice(SHORT_ID_CHARS) for _ in range(length)))


def generate_human_code(length: int = 6) -> str:
    """Generate human-readable code (no ambiguous characters)."""
    return "".join(secrets.choice(HUMAN_CODE_CHARS) for _ in range(length))


# ============================================================================
# UUID UTILITIES
# ============================================================================


def uuid_to_compact(uuid: UUID) -> CompactUUID:
    """Convert UUID to compact form (no dashes)."""
    return CompactUUID(uuid.replace("-", "").lower())


def compact_to_uuid(compact: CompactUUID) -> UUID:
    """Convert compact UUID to standard form."""
    h = compact.lower()
    return UUID(f"{h[:8]}-{h[8:12]}-{h[12:16]}-{h[16:20]}-{h[20:]}")


def uuid_to_bytes(uuid: UUID) -> bytes:
    """Convert UUID to bytes."""
    hex_str = uuid.replace("-", "")
    return bytes.fromhex(hex_str)


def bytes_to_uuid(b: bytes) -> UUID:
    """Convert bytes to UUID."""
    if len(b) != 16:
        raise ValueError("UUID must be 16 bytes")
    h = b.hex()
    return UUID(f"{h[:8]}-{h[8:12]}-{h[12:16]}-{h[16:20]}-{h[20:]}")


# ============================================================================
# ULID UTILITIES
# ============================================================================


def ulid_to_timestamp(ulid: ULID) -> int:
    """Extract timestamp from ULID (milliseconds)."""
    timestamp = 0
    for i in range(10):
        timestamp = timestamp * 32 + ULID_ENCODING.index(ulid[i].upper())
    return timestamp


def ulid_to_datetime(ulid: ULID) -> datetime:
    """Extract datetime from ULID."""
    ts = ulid_to_timestamp(ulid)
    return datetime.fromtimestamp(ts / 1000, tz=timezone.utc)


# ============================================================================
# SNOWFLAKE ID UTILITIES
# ============================================================================

# Discord-style snowflake epoch (2015-01-01)
SNOWFLAKE_EPOCH = 1420070400000


def snowflake_to_timestamp(snowflake: SnowflakeId) -> int:
    """Extract timestamp from Snowflake ID (milliseconds)."""
    return (int(snowflake) >> 22) + SNOWFLAKE_EPOCH


def snowflake_to_datetime(snowflake: SnowflakeId) -> datetime:
    """Extract datetime from Snowflake ID."""
    ts = snowflake_to_timestamp(snowflake)
    return datetime.fromtimestamp(ts / 1000, tz=timezone.utc)


# ============================================================================
# EXPORTS
# ============================================================================

__all__ = [
    # Types
    "UUID",
    "UUIDv7",
    "CompactUUID",
    "ULID",
    "KSUID",
    "NanoID",
    "ShortId",
    "ObjectId",
    "SnowflakeId",
    # Validation
    "is_valid_uuid",
    "is_valid_uuid_v7",
    "is_valid_uuid_any",
    "is_valid_compact_uuid",
    "is_valid_ulid",
    "is_valid_ksuid",
    "is_valid_nano_id",
    "is_valid_short_id",
    "is_valid_human_code",
    "is_valid_object_id",
    "is_valid_snowflake_id",
    "is_valid_ean13",
    "is_valid_upc_a",
    "is_valid_isbn13",
    "is_valid_isbn10",
    "is_valid_doi",
    "is_valid_orcid",
    "is_valid_stripe_customer_id",
    "is_valid_stripe_payment_intent_id",
    "is_valid_stripe_subscription_id",
    "is_valid_arn",
    "is_valid_github_repo",
    "is_valid_k8s_name",
    "is_valid_api_key",
    # Generation
    "generate_uuid",
    "generate_ulid",
    "generate_short_id",
    "generate_human_code",
    # UUID utilities
    "uuid_to_compact",
    "compact_to_uuid",
    "uuid_to_bytes",
    "bytes_to_uuid",
    # ULID utilities
    "ulid_to_timestamp",
    "ulid_to_datetime",
    # Snowflake utilities
    "snowflake_to_timestamp",
    "snowflake_to_datetime",
    # Constants
    "ID_PATTERNS",
    "ULID_ENCODING",
    "SNOWFLAKE_EPOCH",
]
